### PR TITLE
Feature: Add ontology concepts exporters to diffrent formats (xml, ntriples, turtle) 

### PR DESCRIPTION
**Pull request by imadbourouche**
### Description
This feature  enhances the display of ontology concepts by introducing the raw data in a table view for showcasing concept content in multiple formats. This enhancement significantly improves the usability and functionality of the interface by allowing users to easily view concept content in HTML, JSON, XML, NTriples, and Turtle formats.

Screenshot from 2024-03-20 17-28-35

### CODE REVIEW

File: app/javascript/controllers/content_finder_controller.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import { Controller } from '@hotwired/stimulus'
+import hljs from 'highlight.js/lib/core'
+import xml from 'highlight.js/lib/languages/xml'
+import json from 'highlight.js/lib/languages/json'
+
+export default class extends Controller {
```
Comment (syphax-bouazzouni) [inline app/javascript/controllers/content_finder_controller.js:6]: re-use the existent stimulus controllers , don't create a new one

File: app/components/clipboard_component/clipboard_component_controller.js [hunk 2]
```diff
@@ -14,6 +14,21 @@ export default class extends Controller {
       this.#copied()
     })
   }
+  
+  copy_concept_content () {
```
Comment (syphax-bouazzouni) [inline app/components/clipboard_component/clipboard_component_controller.js:18]: no the good way to use this, ask @Bilelkihal

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]

File: app/views/properties/_show.html.haml [hunk 3]
```diff
@@ -13,3 +13,15 @@
           - t.add_row({th: t('properties.preferred_name')}, {td: display_in_multiple_languages(c.concept_properties[:label][:values])}) if  c.concept_properties[:label][:values].present?
           - t.add_row({th: t('properties.definitions')}, {td: display_in_multiple_languages(c.concept_properties[:definition][:values])}) if c.concept_properties[:definition][:values].present?
           - t.add_row({th: t('properties.parent')}, {td: display_in_multiple_languages(c.concept_properties[:parents][:values])}) if c.concept_properties[:parents][:values].present?
+
+      .property-content-format
```
Comment (syphax-bouazzouni) [inline app/views/properties/_show.html.haml:17]: don't have code duplication,
this part is the same in three files, instances, concept_details and this one

File: app/components/clipboard_component/clipboard_component_controller.js (hunk 2 above)
Comment (syphax-bouazzouni) [reply app/components/clipboard_component/clipboard_component_controller.js:18]: still not resolved, this function needs to disappear.

File: app/javascript/controllers/content_finder_controller.js (hunk 1 above)
Comment (syphax-bouazzouni) [reply app/javascript/controllers/content_finder_controller.js:6]: still to do

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]

File: app/views/properties/_show.html.haml [hunk 4]
```diff
@@ -13,3 +13,9 @@
...
```
Comment (syphax-bouazzouni) [inline app/views/properties/_show.html.haml:17]: this block is already in the concept details component that you called up, why do it again here ?

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -629,4 +629,17 @@ def cancel_button_component(class_name: nil, id: , value:, data: nil)
       end
     end
   end
+
+  def content_formats(concept_id: nil, acronym: nil, c: nil, selected_format: nil)
```
Comment (syphax-bouazzouni) [inline app/helpers/application_helper.rb:633]: put it in ontologies_helper.rb file as related to ontologies only  (not for example to users or other models)

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]

File: app/controllers/ontologies_controller.rb [hunk 6]
```diff
@@ -219,6 +219,37 @@ def sparql
     end
   end
 
+  def content_finder
```
Comment (imadbourouche) [inline app/controllers/ontologies_controller.rb:222]: use Faraday

File: app/javascript/controllers/index.js [hunk 7]
```diff
@@ -102,3 +102,6 @@ application.register('form-url', FormUrlController)
 
 import ontologiesSelector from "./ontologies_selector_controller"
 application.register("ontologies-selector", ontologiesSelector)
+
+import ContentFinderController from "./content_finder_controller"
```
Comment (imadbourouche) [inline app/javascript/controllers/index.js:106]: Remove this controller from index.js

File: app/views/ontologies/resource_content.html.haml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+= render TurboFrameComponent.new(id:"resource_content_frame_#{@format}") do
```
Comment (imadbourouche) [inline app/views/ontologies/resource_content.html.haml:1]: do it like this one: https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/511#discussion_r1541989170

File: app/views/properties/_show.html.haml (hunk 4 above)
Comment (imadbourouche) [reply app/views/properties/_show.html.haml:17]: in the concept details, properties and instances, i just took the last part of the component which is the tabcontainer (json,xml,ntriples,turtle), the Raw data (html) part differ from these three

File: app/controllers/ontologies_controller.rb [hunk 9]
```diff
@@ -219,6 +219,37 @@ def sparql
...
+        accept_header = "application/json"
+      when 'xml'
+        accept_header = "application/xml"
+      when 'ntriples'
+        accept_header = "application/n-triples"
+      when 'turtle'
+        accept_header = "text/turtle"
+      end
+
+      url = URI.parse("#{rest_url}/ontologies/#{params[:acronym].strip}/resolve/#{helpers.escape(params[:uri].strip)}")
```
Comment (syphax-bouazzouni) [inline app/controllers/ontologies_controller.rb:237]: create a helper to this in another file ,
this file is already too big and do a lot of things.

File: app/javascript/controllers/index.js [hunk 10]
```diff
@@ -102,3 +102,6 @@ application.register('form-url', FormUrlController)
 
 import ontologiesSelector from "./ontologies_selector_controller"
 application.register("ontologies-selector", ontologiesSelector)
+
+//import ContentFinderController from "./content_finder_controller"
```
Comment (syphax-bouazzouni) [inline app/javascript/controllers/index.js:106]: if not used remove it

File: app/views/properties/_show.html.haml (hunk 4 above)
Comment (syphax-bouazzouni) [reply app/views/properties/_show.html.haml:17]: Your change needs to be only the `concept details component` file.

Properties, Collections, Schemes, use it and it should work without changing each of them.

File: config/routes.rb [hunk 11]
```diff
@@ -122,6 +122,7 @@
   # Ontologies
   get '/ontologies/view/edit/:id' => 'ontologies#edit_view', :constraints => { id: /[^\/?]+/ }
   get '/ontologies/view/new/:id' => 'ontologies#new_view'
+  get '/content_finder' => 'ontologies#content_finder'
```
Comment (syphax-bouazzouni) [inline config/routes.rb:125]: change it to 

get '/ontologies/:acronym//:id/serialize/:format'

File: app/javascript/controllers/metadata_downloader_controller.js [hunk 12]
```diff
@@ -30,6 +31,71 @@ export default class extends Controller {
         hljs.registerLanguage('xml', xml)
         this.showNTriples()
         break
+      case 'ntriples':
+        hljs.registerLanguage('ntriples', function (hljs) {
+          var URL_PATTERN = /<[^>]+>/; // Regex pattern for matching URLs in angle brackets
```
Comment (syphax-bouazzouni) [inline app/javascript/controllers/metadata_downloader_controller.js:36]: URL_PATTERN is not used here

File: app/javascript/controllers/metadata_downloader_controller.js [hunk 13]
```diff
@@ -30,6 +31,71 @@ export default class extends Controller {
...
+              },
+              hljs.COMMENT('^#', '$')
+            ]
+          };
+        });
+        this.showNTriples()
+        break
+      case 'turtle':
+        hljs.registerLanguage('turtle', function (hljs) {
+          var URL_PATTERN = /(?:<[^>]*>)|(?:https?:\/\/[^\s]+)/;
```
Comment (syphax-bouazzouni) [inline app/javascript/controllers/metadata_downloader_controller.js:60]: use `let` not `var`

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]

**Review by syphax-bouazzouni — APPROVED** [review]

**Pull request merged**